rllib/core/learner/learner_group.py

```diff
@@ -231,43 +239,77 @@ def _distributed_update(
+            A list of dictionaries of results from the updates from the individual
+            Learner(s)
         """
+        # Make sure minibatch size is reduced to the correct number of shards as well
+        # (just like we split each batch into the number of learner workers).
+        if minibatch_size is not None:
+            minibatch_size //= len(self._workers)
+
+        def _learner_update(learner, minibatch):
+            return learner.update(
+                minibatch,
+                minibatch_size=minibatch_size,
+                num_iters=num_iters,
+                reduce_fn=reduce_fn,
+            )
 
         if block:
-            results = self._worker_manager.foreach_actor(
-                [
-                    lambda w: w.update(
-                        b,
-                        minibatch_size=minibatch_size,
-                        num_iters=num_iters,
-                        reduce_fn=reduce_fn,
-                    )
-                    for b in ShardBatchIterator(batch, len(self._workers))
-                ]
-            )
-        else:
-            if batch is not None:
-                self._in_queue.append(batch)
-            results = self._worker_manager.fetch_ready_async_reqs()
-            if self._worker_manager_ready() and self._in_queue:
-                batch = self._in_queue.popleft()
-                self._worker_manager.foreach_actor_async(
+            results = self._get_results(
+                self._worker_manager.foreach_actor(
                     [
-                        lambda w: w.update(
-                            b,
-                            minibatch_size=minibatch_size,
-                            num_iters=num_iters,
-                            reduce_fn=reduce_fn,
-                        )
-                        for b in ShardBatchIterator(batch, len(self._workers))
+                        partial(_learner_update, minibatch=minibatch)
+                        for minibatch in ShardBatchIterator(batch, len(self._workers))
                     ]
                 )
+            )
+        else:
+            # Queue the new batches.
+            # If queue is full, kick out the oldest item (and thus add its
+            # length to the "dropped ts" counter).
+            if len(self._in_queue) == self._in_queue.maxlen:
+                self._in_queue_ts_dropped += len(self._in_queue[0])
+
+            self._in_queue.append(batch)
+
+            # Retrieve all ready results (kicked off by prior calls to this method).
+            results = self._worker_manager.fetch_ready_async_reqs()
+            # Only if there are no more requests in-flight on any of the learners,
+            # we can send in one new batch for sharding and parallel learning.
+            if self._worker_manager_ready():
+                count = 0
+                # TODO (sven): This probably works even without any restriction
```

Comment:
Again, will try max_in_flight==3 first, but if it turns out that it doesn't matter (and no restriction works as well), we can remove this check entirely and might even remove the internal queue (always process everything that comes in, no matter what).